app/src/main/java/de/tum/in/tumcampusapp/activities/RoomFinderActivity.java

```diff
@@ -64,92 +66,79 @@ public void onCreate(Bundle savedInstanceState) {
     }
 
     @Override
-    protected void onStartSearch() {
-        List<Map<String, String>> recents = getRecents();
-        if (recents.isEmpty()) {
-            finish();
-            return;
-        }
-        adapter = new RoomFinderListAdapter(this, recents);
-        list.setAdapter(adapter);
+    protected Optional<List<RoomFinderRoom>> onSearchInBackground() {
+        return Optional.of(getRecents());
     }
 
     @Override
-    public void onStartSearch(String query) {
-        roomFinderRequest.fetchSearchInteractive(this, this, query);
+    protected Optional<List<RoomFinderRoom>> onSearchInBackground(String query) {
+        try {
+            List<RoomFinderRoom> rooms = TUMCabeClient.getInstance(this).fetchRooms(query);
+            return Optional.of(rooms);
+        } catch (IOException e) {
+            Utils.log(e);
+        }
+        return null;
     }
 
     @Override
-    public void onFetch(List<Map<String, String>> result) {
-        if (result.isEmpty()) {
+    protected void onSearchFinished(Optional<List<RoomFinderRoom>> result) {
+        if(!result.isPresent()){
+            if (NetUtils.isConnected(this)) {
+                showErrorLayout();
+            } else {
+                showNoInternetLayout();
+            }
+            return;
```

Comment:
`return; ` is necessary because there is no result to display. Otherwise it will cause `NullPointerException` .